Elevation behavior

podraco opened the pull request:
### Description of Changes
Lastest set of updates for the Real time Elevation behavior with mockups added.

### Screenshots

https://user-images.githubusercontent.com/5509325/107318042-6c859d80-6a61-11eb-972d-815382a45ec3.mp4


https://user-images.githubusercontent.com/5509325/107318046-6db6ca80-6a61-11eb-940d-075f60ddc799.mp4

podraco committed:
Elevation update

Manuel Ortiz

- #794

Shadow generation in real time, it will depend on of the widget size and
shape.

RoundedRectangularElevationBehavior class added, this class is in
charge to allow developer to draw rounded rectangles efficiently for
both shadows.

CommonElevationBehavior is now in charge of every math process that is
shared between different shadows.

1. "RectangularElevationBehavior"
2. "CircularElevationBehavior"
3. "RoundedRectangularElevationBehavior"

Larger widgets will cast a bigger shadow, thus a bigger texture.

Added Rotation support:
 1. Hard shadow shares origin with the widget
 2. Soft shadow has his own origin. Improving the mock up of a real
   shadow work better.

Fake light sourcess support added

Shadow Groups added

Canvas sort support added. (canvas instruction can be found by group
inside `canvas.before`)

To set a new shadow draw function, set the function inside _draw_shadow.
This allows developers to use external functions to draw the shadow
over the canvas.

BackgroundColorBehavior now has rotation support. this allows
independent rotation in order to use:
```py
canvas:
    PushMatrix:
    Trans:
    instr:
    instr:
    instr:
    PopMatrix:
```
Instead of stacking the whole canvas inside the matrix.

DO NOT STACK EVERYTHING FORM DIFFERENT LAYERS.
For example:
```py
canvas.before:
    PushMatrix:
    Trans:
    instr:
    instr:
    instr:
canvas:
    PopMatrix:
```
Is now forbidden because this means that it will take an existing
transformation of the canvas and over do it,  for example, if it's
supposed to have a determined rotation, it will rotate with the new
transformation and then again with the original transformation.
Thus generating visual errors.

Angle is now a reserved word of the background and elevation behaviors
class.
Both classes share the same property and acts independently.

Radius is now a VariableListProperty, this allows an easy set to
every developer.

Cambios a ser confirmados:
	Modificado:     kivymd/uix/behaviors/backgroundcolor_behavior.py
	modificado:     kivymd/uix/behaviors/elevation.py

RoundedRectangularElevationBehavior added to uix/behaviors/__init__.py

Fixed Missing bind to _elevation.

Update documentation of elevation behavior

podraco committed:
Elevation Kitchen sink Demo

The new elevation class and shaders is now documented and added to the 
kitchen sink demo app

podraco committed:
Included improved old shadow definitions.

The old clases for shadow casting are back, but now are called 
FakeRectangularElevationBehavior and FakeCircularElevationBehavior 

Size and transparency are now fixed
The index of the elevation can accept a higher index than 23 (the last 
shadow in the atlas for shadows in both classes)

podraco committed:
bugfix correction

Navigation drawer was taking control from the elevation behavior class, 
causing some bugs in screen and calling re-draws when is unnecesary

podraco committed:
Bug Fix cards now draw with the correct shadow class

podraco committed:
Bugfix and corrections

The BaseElevationButton class was not respecting the Material IO 
standard.
It states in it's examples that every elevated button shoud have 6 
levels of elevation between it's resting state and its raised state.

Bugfix:
The button class was taking control form the elevation behavior class 
thus re drawing the button multiple times.

podraco committed:
Added new controls in the demo screen

This controls are meant to allow users to play with the widgets visual 
properties.
Radius and elevation.

podraco committed:
Improved Shadows and layout for the demo

podraco committed:
Updated docs and values

Included a section on how to properly animate the elevation of the 
widget with the elevaiton behavior.

some bug fixes and optimizations.

podraco committed:
Update of the elevatoin demos

Included the fake shadows to the demo screen

Support for a future upgrade included, not implemented yet.

podraco committed:
Removed development flag

podraco committed:
Improved performance.

podraco committed:
Pre-commit run commit

HeaTTheatR commented:
@podraco 

![Снимок экрана 2021-02-09 в 08 22 39](https://user-images.githubusercontent.com/16930280/107319494-01af8300-6ab0-11eb-81fe-e3b7ea307051.png)

`demos/kitchen_sink/libs/baseclass/elevation.py`
`demos/kitchen_sink/libs/kv/elevation.kv`
`demos/kitchen_sink/screens_data.json`

**You didn't need to modify these files.**

These examples overload the `KitchenSink` application. Therefore, I want to put them in the documentation. For demonstration, it is enough that there is now:


https://user-images.githubusercontent.com/16930280/107319789-95814f00-6ab0-11eb-9103-4ad14f85db08.mov

podraco commented:
Ok then. Should i delete all the modifications i did to the demo mods then?

HeaTTheatR commented:
@podraco You just haven't done a `git pull` before making new changes :)

podraco commented:
i'm reviewing the files...

HeaTTheatR commented:
@podraco You need to do a `git pull`, make changes, and then do a `git push`

podraco commented:
@HeaTTheatR , the master branch didn't allow me to work directly with it, i managed to do it by a fork repo that was up to date with the master branch when i started doing changes.
Let me review the files and ill discard whatever is not needed by my end.

HeaTTheatR commented:
@podraco OK.

podraco committed:
Merge branch with master update

Both files have been merged to add the spell check that HeaTTheatR did.

Thanks for your hard work bro, and sorry.

podraco committed:
Merged Demo with master

podraco committed:
Merge branch 'master' into master

podraco committed:
Pre-commit run

podraco committed:
Merge branch 'master' of github.com:podraco/KivyMD

HeaTTheatR commented on kivymd/uix/behaviors/elevation.py:
```diff
@@ -1145,37 +1145,25 @@ def __init__(self, **kwargs):
-            img = img.filter(
-                ImageFilter.GaussianBlur(
-                    radius=int(dp(1 + self.hard_shadow_offset / 3))
+            # Set shadow size.
+            ratio = self.width / (self.height if self.height != 0 else 1)
+            if -2 < ratio < 2:
+                self._shadow = MDApp.get_running_app().theme_cls.quad_shadow
+                width = soft_width = self.width * 1.9
+                height = soft_height = self.height * 1.9
+            elif ratio <= -2:
+                self._shadow = MDApp.get_running_app().theme_cls.rec_st_shadow
+                ratio = abs(ratio)
+                if ratio > 5:
+                    ratio = ratio * 22
+                else:
+                    ratio = ratio * 11.5
+                width = soft_width = self.width * 1.9
+                height = self.height + dp(ratio)
+                soft_height = (
+                    self.height + dp(ratio) + dp(self._elevation) * 0.5
```
Are we back to old changes?

podraco replied:
more like mockup shadows for bigger elements, 
the new shadows render the background dynamically, the old draws a mockup texture.
while the new shadows cast a better and 
stylized shadow and elevation. you can see how the shadows differ between the Fake**elevationbehavior and the Rectangular/circularElevarion behavior in the videos above.

podraco replied:
While material IO ask for a dynamic shadow cast, stating that shadow should change with the widget, fake shadows are a great tool for most static widgets.

HeaTTheatR commented:
@podraco Please clarify the status of this pull request at the moment, because I'm already confused :))

podraco commented:
The pull request divides on 2 parts.

#### First part | Elevation.py.

These are the elevation classes that are currently available in the behavior

##### Real-time Shadows:
1. `RectangularElevationBehavior`: Draws a rectangular shadow based on the size of the widget that keeps consistency and keeps a natural-looking shadow
2. `CircularElevationBehavior`:  Draws a Circular shadow based on the size of the widget that keeps consistency and keeps a natural-looking shadow
3. `RoundedRectangularElevationBehavior`: Draws a rounded cornered rectangular shadow based on the size of the widget that keeps consistency and keeps a natural-looking shadow
4. `ObservableShadow`: it's  a proposal to allow faking partial shadows that I'm currently working on, this is supposed to draw only the part of the shadow that's on screen. - Still under development, but the name is already added for future use.


##### Fake Shadows:
1. `FakeRectangularElevationBehavior`: Fixed shadow based on preprocessed rectangular and square textures, the old way,
2. `FakeCircularElevationBehavior`: Fixed shadow based on preprocessed Circular textures, the old way,

The problem with this is that the shadow image and saturation won't match between a rectangular shadow and a square shadow. It's faster because uses stored shadows to draw, but it's not as clean as the Real-time elevation classes. 

This part of the pull request adds this classes back for developers to decide which elevation behavior their implementation requires.
While keeping the real-time shadows as the standard for every prefab widget.

This part of the pull request is intended for flexibility of kivymd.


#### Second part | Optimizations.
The second part of this pull request adds some optimizations, taking away the elevation control that some widgets took form the behavior, while adding a new section to the elevation behavior docs on how to properly animate the elevation of the widgets affected.

HeaTTheatR commented:
@podraco But I still don't see any documentation on how to use ...

podraco commented:
the idea is that they use them as any other behavior, all processing is internally done so they can just change the behavior they need, like 
``` py
class someclass(
    themable_behavior,
    shadow_behavior, # here they can choose between all our behaviors and they will work with the same standard API
    Specific_Color_Bheavior,
    ripple_behavior,
    higher_class_behavior.
):
```

HeaTTheatR commented:
@podraco Unfortunately, I do not understand what this is about ...

podraco commented:
i'm sorry i'm not sure how to express it more clearly.

HeaTTheatR commented:
@podraco Me and users need specific code examples ...

gregynard commented:
FYI the elevation for card is okay but the MDMDFloatingActionButton still the same. The resolution also is kinda different from the previous. the program is still sluggish there are some delays when you pressed the button and switch screens. 
Thanks!

HeaTTheatR commented:
@gregynard Sorry, I know there are big problems with the new changes that related  this pull request. Unfortunately, the author of this pull request has not yet provided documentation on how to properly use these changes. Let's wait ...

gregynard commented:
Oki oki. No problem. Thanks!

podraco commented:
Sorry I was testing a theory. It seems that it worked :9
in the next pull request ill provide the necessary docs.

HeaTTheatR commented:
@podraco 

Test with Pillow:


https://user-images.githubusercontent.com/16930280/107553682-0e35f780-6be6-11eb-9bd5-d4246b7d6698.mov

Test with Pillow-SIMD:


https://user-images.githubusercontent.com/16930280/107553756-2279f480-6be6-11eb-8f3e-52c11e226cc4.mov

Pillow-Simd - https://github.com/uploadcare/pillow-simd

podraco commented:
never heard of that i was attepmting to use multithread to cast shadows in a secondary process, since the actual one blocks the GIL

podraco commented:
how well does the corners works?

gregynard commented:
Hi @podraco the elevations are now okay. But there are still some delays when switching screen or expanding a card. The mdfloating button also has no elevation. Thanks!

podraco commented:
hii @gregynard 
the elevation documentation has been updated, i recommend you to check it up.
the new elevation classes uses real-time shadows, since the Material IO requires that the casted shadow changes with the widget, but, the FakeRectangular and FakeCircular elevation behaviors can improve the rendering time on square cards. and circular elements.
You can compile the latest docs directly with a local clone of the git repo or read the file directly.
as for the MDFloating button, i had no idea, let me check it up.

gregynard commented:
Hi @podraco this is noted. Thank you!

podraco commented:
got a fix for all buttons. but it seems the shadow of the MDToolbar MDfloating action button was broken all along...

gregynard commented:
Hi @podraco thanks for this. I tried to follow the documentations I can now control the elevations.

podraco commented:
it depends on how kivy is installed and used, as you can see, @HeaTTheatR  used Pillow-SIMD which shown a great performance improvement.
i as well am working on some performance solutions that will keep the shadows light weight and faster, but for now, I'm looking through the code, since some widgets take control from the behavior and duplicate some actions, like, redrawing the shadow multiple times outside the behavior or require a mock-up shadow instead of an RTC

podraco commented:
If you're interested in some previews of the solutions, you're always welcome to checkup podraco/kivymd branches, I would appreciate some feedback before requesting a merge.

gregynard commented:
Noted on this. Thanks thanks!

podraco commented:
the last commits are now part of the master branch
